[libevent] update to 2.1.12

DailyShana opened the pull request:
**Describe the pull request**

- What does your PR fix? Fixes #
update libevent to 2.1.12

- Which triplets are supported/not supported? Have you updated the CI baseline?
no change

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

DailyShana committed:
[libevent] update to 2.1.12

DailyShana committed:
fix path in LibeventConfig.cmake.in

DailyShana commented:
for port `folly` most errors are like this
```
CMake Error at D:/installed/x64-windows-static/share/libevent/LibeventTargets-static.cmake:81 (message):
  The imported target "libevent::core" references the file

     "D:/packages/libevent_x64-windows-static/debug/lib/event_core.lib"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "D:/installed/x64-windows-static/share/libevent/LibeventTargets-static.cmake"

  but not all the files it references.
```
It tried to find the lib in package dir.

but I can install `folly:x64-windows-static` successfully on my machine, found libevent normally:
```
-- Found libevent: D:/vcpkg/installed/x64-windows-static/debug/lib/event.lib
```
in the installed dir

DailyShana committed:
configure crt runtime

NancyLi1013 commented:
/azp run

NancyLi1013 commented:
@BillyONeal 
I cannot reproduce this on my machine either. It seems there is no `D:/installed/x64-windows-static/share/libevent/LibeventTargets-static.cmake`. The file should be `LibeventTargets.cmake`.
![image](https://user-images.githubusercontent.com/46708020/87020763-e7a27f00-c206-11ea-9940-ac203a8699b7.png)


Could you please help take a look?

DailyShana commented:
I can reproduce if I just remove the libevent package from `vcpkg_root/packages`

NancyLi1013 commented:
@DailyShana 
Could you please look into this?

DailyShana committed:
patch target file

DailyShana committed:
try to fix evpp build on unix

DailyShana committed:
try to fix evpp build on unix

DailyShana commented:
`folly` has no problem now.

`fizz:x64-windows` failed because it cannot find libevent but `fizz:x64-windows-static` works normally
I can reproduce it but really can't understand it

`evpp` failed on osx and linux because the linker cannot find the libevent libary. Because I have no condition to build on osx or linux, I just tried to update the patch but it seems useless.

DailyShana committed:
try to fix evpp build on unix

NancyLi1013 commented:
/azp run

NancyLi1013 commented:
Hi @DailyShana 
For `evpp`, there is no need to add` fix-linux-build.pach`. I removed this patch and built on my local, it failed  with this:

```
/home/vruil/12277/vcpkg/buildtrees/evpp/src/v0.7.0-5b8db028cd/evpp/logging.h:9:10: fatal error: glog/logging.h: No such file or directory
    9 | #include <glog/logging.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
```
You might need to check if the header path is set.

For `fizz`, I also don't know why it cannot be built successfully with dynamic. But it can indicate that there is something wrong with `libevent ` or its usage in `fizz`. The components are not found now.
Could you please check this again?

DailyShana committed:
try to fix evpp build on unix

DailyShana commented:
`fix-linux-build.patch` is necessary, it solves the problem of finding glog headers by using targets `glog::glog`

I modified the patch to use imported targets `libevent::core` `libevent::extra` `libevent::openssl` but it has the following errors:
```
  Target "evpp" links to target "-lpthreadlibevent::core" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  evpp/CMakeLists.txt:23 (add_library)
```

```
CMake Error in evpp/CMakeLists.txt:
  Imported target "libevent::openssl" includes non-existent path

    "/mnt/vcpkg-ci/installed/x64-linux/include/mnt/vcpkg-ci/installed/x64-linux/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```
maybe there is something wrong with the imported targets of libevent on linux, but I don't have linux or osx machine to test it.

DailyShana commented on ports/libevent/portfile.cmake:
```diff
@@ -38,20 +42,27 @@ vcpkg_configure_cmake(
+else()
+    set(_target_suffix shared)
+endif()
+file(READ ${CURRENT_PACKAGES_DIR}/share/libevent/LibeventTargets-${_target_suffix}.cmake _content)
+string(REPLACE "${CURRENT_PACKAGES_DIR}" "${CURRENT_INSTALLED_DIR}" _content ${_content})
+file(WRITE ${CURRENT_PACKAGES_DIR}/share/libevent/LibeventTargets-${_target_suffix}.cmake ${_content})
```
Weirdly, after reading and writing, all semicolons `;` in the file disappeared

DailyShana committed:
fix file patching

NancyLi1013 commented on ports/libevent/portfile.cmake:
```diff
@@ -38,20 +42,27 @@ vcpkg_configure_cmake(
+else()
+    set(_target_suffix shared)
+endif()
+file(READ ${CURRENT_PACKAGES_DIR}/share/libevent/LibeventTargets-${_target_suffix}.cmake _content)
+string(REPLACE "${CURRENT_PACKAGES_DIR}" "${CURRENT_INSTALLED_DIR}" _content "${_content}")
+file(WRITE ${CURRENT_PACKAGES_DIR}/share/libevent/LibeventTargets-${_target_suffix}.cmake "${_content}")
```
```suggestion
vcpkg_replace_string(${CURRENT_PACKAGES_DIR}/share/libevent/LibeventTargets-${_target_suffix}.cmake
 "${CURRENT_PACKAGES_DIR}" 
 "${CURRENT_INSTALLED_DIR}"
  )
```
Could you please use `vcpkg_replace_string()` to replace a string in a file?

NancyLi1013 reviewed (commented):
Could you please also bump the version in _ports/evpp/CONTROL_?

DailyShana committed:
code review

DailyShana committed:
fix fizz:x64-windows build

DailyShana committed:
Merge branch 'microsoft/master' into libevent

NancyLi1013 commented on ports/fizz/portfile.cmake:
```diff
@@ -3,18 +3,16 @@ vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
 # Prefer installed config files
 file(REMOVE
     ${SOURCE_PATH}/fizz/cmake/FindGflags.cmake
-    ${SOURCE_PATH}/fizz/cmake/FindLibevent.cmake
```
This file seems to use module mode to find `libevent`. So we should still remove this file and try to find `libevent` via config mode.

DailyShana replied:
It works very well, why remove it?

NancyLi1013 replied:
Since this file is provided by `fizz `instead of vcpkg. We prefer to use xxxconfig.cmake to find the package, which is better to maintain.

DailyShana replied:
But the fact is removing it makes harder to maintain, one have to patch many things.
I think if the port can find the dependences properly, we don't need to change it, no matter it is config mode or module mode.

NancyLi1013 commented:
/azp run

strega-nil commented:
LGTM, thanks @DailyShana :)
